packages/periphery/contracts/oracles/PendleOracle.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity 0.8.19;
+
+import '@openzeppelin/contracts/access/Ownable2Step.sol';
+import '@openzeppelin/contracts/utils/math/Math.sol';
+import '@pendle/core-v2/contracts/oracles/PendlePtLpOracle.sol';
+import '@pendle/core-v2/contracts/core/Market/v3/PendleMarketV3.sol';
+import '@marginly/contracts/contracts/interfaces/IPriceOracle.sol';
+
+/// @notice Oracle for Pendle PT / underlying token, e.g. PT-ezETH-27JUN2024 / WETH
+contract PendleOracle is IPriceOracle, Ownable2Step {
+  struct OracleParams {
+    address pendleMarket;
+    uint16 secondsAgo;
+    uint16 secondsAgoLiquidation;
+    address yieldToken;
+    address secondaryPoolOracle;
+    uint8 ptDecimals;
+    uint8 syDecimals;
+    bool usePtToSyOracle;
+  }
+
+  error ZeroPrice();
+  error ZeroAddress();
+  error WrongValue();
+  error PendlePtLpOracleIsNotInitialized(uint16);
+
+  IPPtLpOracle public immutable pendle;
+  mapping(address => mapping(address => OracleParams)) public getParams;
+
+  uint256 private constant X96ONE = 2 ** 96;
+  uint256 private constant PRICE_DECIMALS = 18;
+
+  constructor(address _pendle) {
+    if (_pendle == address(0)) revert ZeroAddress();
+    pendle = IPPtLpOracle(_pendle);
+  }
+
+  function setPair(
+    address quoteToken,
+    address baseToken,
+    address pendleMarket,
+    address secondaryPoolOracle,
+    address yieldToken,
+    uint16 secondsAgo,
+    uint16 secondsAgoLiquidation,
+    bool usePtToSyOracle
+  ) external onlyOwner {
+    if (secondsAgo == 0 || secondsAgoLiquidation == 0) revert WrongValue();
+    if (
+      quoteToken == address(0) ||
+      baseToken == address(0) ||
+      pendleMarket == address(0) ||
+      secondaryPoolOracle == address(0) ||
+      yieldToken == address(0)
+    ) {
+      revert ZeroAddress();
+    }
+
+    _assertOracleIsInitialized(pendleMarket, secondsAgo);
+    _assertOracleIsInitialized(pendleMarket, secondsAgoLiquidation);
```

Comment:
I would guess the `secondsAgo` check is sufficient enough: `secondsAgoLiquidation` has lower value and should always pass if the previous one has succeeded. However some verification of my statement is required